### PR TITLE
Implement Publish-Module credential parameter propagation to sub-functions

### LINKED ISSUE #103: Publish-Module does not propagate Credentials to sub-functions
Publish-Module cannot verify if Dependant Modules are already present in cases where PSReposity should be accessed with alternate credential.

## Expected Behavior
If Publish-Module is been run with "Credential" parameter it should propagate the provided Credentials to the other functions that are doing validations against the PSRepository.
The suggested solution is to detect if "Credential" parameter is being supplied and propagate it to the sub-functions.

## Current Behavior
When you try to publish a module that has dependencies defined in it`s ".psd1" file it fails to properly detect if they are already present on the PSRepository when the following conditions are present:
- Alternative credentials are being provided using the "Credential" parameter
- The Credentials supplied are different from the credentials of the user running the process.

## Possible Solution
Implement optional Credential parameter to the following functions:
- Publish-PSArtifactUtility
- Get-ModuleDependencies
- ValidateAndGet-RequiredModuleDetails

and detect and propagate the Credential parameter if it have been supplied.

## Steps to Reproduce (for bugs)
1. Publish-Module -Name testmodule -Repository LocalPsRepo -NuGetApiKey 123 -Credential (Get-Credential)
2. Provide Credentials different from the one used to run the PowerShell process.

## Your Environment

> $PSVersionTable

Name                           Value                                                                                                       
----                           -----                                                                                                       
PSVersion                      5.0.10586.117                                                                                               
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                     
BuildVersion                   10.0.10586.117                                                                                              
CLRVersion                     4.0.30319.42000                                                                                             
WSManStackVersion              3.0                                                                                                         
PSRemotingProtocolVersion      2.3                                                                                                         
SerializationVersion           1.1.0.1                                                                                                     

> Get-Module -ListAvailable PowerShellGet,PackageManagement

ModuleType Version    Name                                ExportedCommands                                                                 
---------- -------    ----                                ----------------                                                                 
Script     1.1.3.0    PackageManagement                   {Find-Package, Get-Package, Get-PackageProvider, Get-PackageSource...}           
Binary     1.0.0.1    PackageManagement                   {Find-Package, Get-Package, Get-PackageProvider, Get-PackageSource...}           
Script     1.1.2.0    PowerShellGet                       {Install-Module, Find-Module, Save-Module, Update-Module...}                     
Script     1.0.0.1    PowerShellGet                       {Install-Module, Find-Module, Save-Module, Update-Module...}                     

> Get-PackageProvider

> Get-PackageProvider -ListAvailable
```

### PR DESCRIPTION
**Pull request by gogbg**
I want to resolve #103 by:

Add Optional Credential Parameter to the following functions:
- Publish-PSArtifactUtility
- Get-ModuleDependencies
- ValidateAndGet-RequiredModuleDetails
and detect and propaget the Credential parameter if it is been used down
the following path:
Publish-Module -> Publish-PSArtifactUtility -> Get-ModuleDependencies ->
ValidateAndGet-RequiredModuleDetails -> Find-Module

### CODE REVIEW

File: PowerShellGet/PSModule.psm1 [hunk 1]
```diff
@@ -7619,11 +7631,7 @@ function ValidateAndGet-ScriptDependencies
         [parameter(Mandatory = $true)]
         [ValidateNotNullOrEmpty()]
         [System.Management.Automation.PSCmdlet]
-        $CallerPSCmdlet,
-
-        [Parameter()]
-        [PSCredential]
-        $Credential
```
Comment (bmanikm) [inline PowerShellGet/PSModule.psm1]: Could you please let me know the reason for removing Credential param on this function? #Closed

File: PowerShellGet/PowerShellGet.psd1 [hunk 2]
```diff
@@ -1,6 +1,6 @@
 ﻿@{
 RootModule = 'PSModule.psm1'
-ModuleVersion = '1.1.3.0'
+ModuleVersion = '1.1.3.1'
```
Comment (bmanikm) [inline PowerShellGet/PowerShellGet.psd1:3]: >'1.1.3.1' [](start = 16, length = 9)

Is it required to bump-up the module version in every PR? #Closed

Comment (gogbg) [reply PowerShellGet/PowerShellGet.psd1:3]: I`m using the module meanwhile, and want to differentiate it between current 1.1.3.0 version #Closed

File: PowerShellGet/PSModule.psm1 (hunk 1 above)
Comment (gogbg) [reply PowerShellGet/PSModule.psm1]: It was unintentional, I`ve reverted this. #Closed

File: PowerShellGet/PSModule.psm1 [hunk 3]
```diff
@@ -7638,7 +7646,6 @@ function ValidateAndGet-ScriptDependencies
 
             $FindModuleArguments = @{
                                         Repository = $Repository
-                                        Credential = $Credential
```
Comment (bmanikm) [inline PowerShellGet/PSModule.psm1:7641]: >Credential = $Credential [](start = 40, length = 24)

Please add corresponding changes here.

            if ($PSBoundParameters.ContainsKey('Credential'))
            {
                $FindModuleArguments.Add('Credential',$Credential)
            } #Closed

File: PowerShellGet/PSModule.psm1 [hunk 4]
```diff
@@ -7703,7 +7710,6 @@ function ValidateAndGet-ScriptDependencies
         {
             $FindScriptArguments = @{
                                         Repository = $Repository
-                                        Credential = $Credential
```
Comment (bmanikm) [inline PowerShellGet/PSModule.psm1:7706]: >Credential = $Credential [](start = 40, length = 24)

Please add corresponding changes here too.

            if ($PSBoundParameters.ContainsKey('Credential'))
            {
                $FindScriptArguments.Add('Credential',$Credential)
            } #Closed

**Comment by bmanikm** [comment]
:clock1:

**Comment by dotps1** [comment]
@bmanikm I thought I already made all these changes in https://github.com/PowerShell/PowerShellGet/pull/93.

**Comment by bmanikm** [comment]
@dotps1 Yes, you are correct, your PR resolved the actual issue. I thought, @gogbg is refactoring those changes.
@gogbg, would you like to close this PR or proceed with refactoring?

File: PowerShellGet/PSModule.psm1 (hunk 3 above)
Comment (gogbg) [reply PowerShellGet/PSModule.psm1:7641]: done with a commit d966927 #Closed

File: PowerShellGet/PSModule.psm1 (hunk 4 above)
Comment (gogbg) [reply PowerShellGet/PSModule.psm1:7706]: should be done with a commit d966927 #Closed

File: PowerShellGet/PSModule.psm1 [hunk 5]
```diff
@@ -8222,7 +8258,6 @@ function Publish-PSArtifactUtility
         $DependentModuleDetails += ValidateAndGet-ScriptDependencies -Repository $Repository `
                                                                      -DependentScriptInfo $PSScriptInfo `
                                                                      -CallerPSCmdlet $PSCmdlet `
-                                                                     -Credential $Credential `
                                                                      -Verbose:$VerbosePreference `
```
Comment (bmanikm) [inline PowerShellGet/PSModule.psm1]: Similar changes are required here too. #Closed

**Comment by gogbg** [comment]
I should be done with refactoring

**Pull request merged**